% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\usepackage{natbib}
\bibliographystyle{abbrvnat}
\setcitestyle{authoryear, open={((},close={)}}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Hypoxia modulates tumor progression by controlling cell cycle proliferation and immunogenicity at the same time},
  pdfauthor={Gepoliano Chaves},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{Hypoxia modulates tumor progression by controlling cell cycle
proliferation and immunogenicity at the same time}
\author{Gepoliano Chaves}
\date{June 19, 2024}

\begin{document}
\maketitle

\hypertarget{introduction}{%
\subsection{INTRODUCTION}\label{introduction}}

Neuroblastoma is a pediatric cancer of the peripheral nervous system.
Due to abnormal cell division rates, cancer cells increase oxygen
consumption and decrease its availability at the tumor microenvironment.
In neuroblastoma, hypoxia is thought to induce dedifferentiation, a
phenomenon thought to drive tumor aggressiveness (Jogi 2002). Therefore,
great interest is involved in studying mechanisms underlying
hypoxia-driven aggressiveness in neuroblastoma.

The biological mechanisms mediating tumor aggressiveness in
neuroblastoma are not completely understood. Pediatric cancers are
thought to be driven by epigenetics rather than tumor mutation. Our
group has investigated an epigenetic mechanism mediated by TET1 enzyme
and deposition of 5-hydroxymethyl-cytosine (5-hmC), a marker of gene
expression, identified in neuroblastoma cells exposed to hypoxia
(Mariani 2014; Hains 2022). The hypoxia-induced TET1 deposition of 5-hmC
was suggested to characterize a cell line transition or plasticity
mechanism, triggered by hypoxia and impacting the two main neuroblastoma
cell populations: ADRN and MES cells. Specifically, it was hypothesized
that hypoxia mediates ADRN to MES (AMT), but not MES to ADRN transition
(Chaves 2023, in preparation).

ADRN cells have neuronal-like features, are faster-growing and comprise
the bulk of tumors, whereas MES cells are stem-like and thought to
associate with resistance to chemotherapy and immunotherapy. It was
suggested that hypoxia causes a change in the biological identity of the
ADRN cell population using 5-hmC deposition, transitioning to a more MES
state maintained under hypoxia (Chaves 2023, in preparation). It is
accepted that MES cells are responsible for mechanisms of minimal
residual disease and relapse of neuroblastoma (cite) after chemotherapy.
Therefore, hypoxia represents an important mechanism mediating cell
transition that leads to metastasis.

The MES cellular state induced by hypoxia caused global and local 5-hmC
depositions on MES genes of the proinflammatory TNFA pathway, notably
the STC1 gene. Surprisingly, this change paradoxically correlated with
both improved overall survival and worse histology (TARGET dataset 149
samples). This information was paradoxical because the MES genes and the
highlighted 5-hmC deposition peaks in the example of STC1 gene, present
in the TNFA pathway had such correlation with positive and negative
outcomes. This has caused us to be interested in investigating the
correlation between HIF1A and outcome in different neuroblastoma (the
Discovery and Validation) cohorts.

Our findings correlate the hypoxia state with the MES cell state
indicative of an inflammatory state, showing that the inflammatory state
is reached with the participation of hypoxia and/or HIF1A and the other
hypoxia transcription factors. Furthermore, it was suggested that
hypoxia induces TNFA signaling via NFKB (Chaves 2023, in preparation),
an important process for inflammation (Liu 2017).

Hypothesizing the existence of a time-dependent relationship between
hypoxia/MES gene set, it may be possible to understand the relationship
between MES-ness and hypoxia over time, as a function of risk (the
relationship of time and risk are not clear in this statement: check).
The MES cell state, while highly related to the immunological phenomenon
of ADCC triggered by dinutuximab according to Mabe 2022, renders MES
cells unable to express GD2, important in dinutuximab immunotherapy.

At the same time, HIF1A is thought to be associated with T-cell
exhaustion (Baldominos 2022) and fundamental for cell-cycle arrest under
hypoxia (Goda, 2003). It was suggested that cancer cells in the
quiescent, non-proliferative state, have reduced antigen in the cell
surface, a phenotype controlled by HIF1A (Baldominos 2022). Tumors that
loose targeted antigens do not need any other means to escape from T
cell attack (Baldominos 2022). Emens et al., (2021) and Baldominos et
al., (2022) suggested cell proliferation, a surrogate for HIF1A
inhibition and cellular differentiation, to be a marker of patients that
responded to ICB. Furthermore, an interplay between c-Myc and HIF1A that
maintains cell-cycle arrest (Druker 2021), via increased p21 levels
(Koshiji 200X), where HIF1A controls the highly demanding need for
energy for the cell cycle to happen (Druker 2021). Therefore, from this
information, it follows that hypoxia and HIF1A have a deep impact on
immunological mechanisms in neuroblastoma, contributing to tumor
progression.

Altered proliferation, increased DNA damage repair mechanisms and
reduced apoptosis are hallmarks of anti-cancer drug resistance (Cree and
Charlton 2017). Cell adhesion also has a role in chemotherapy resistance
(Alimbetov 2018). Consider the Von Groningen 2017 paper for the MES cell
enrichment upon chemotherapy treatment and potential involvement with
hypoxia.

These findings caused us to propose a model where genes modified by
hypoxia have a role on cell wall permeability of blood vessels as part
of the inflammatory process, potentially influenced by time. This is a
hypothesis and we have not yet conducted an experiment to follow-up on
this hypothesis.

\hypertarget{objectives}{%
\subsection{2. OBJECTIVES}\label{objectives}}

\hypertarget{general}{%
\subsubsection{2.1. General}\label{general}}

Our goal is to develop precision oncology algorithms to quantify
phenotypes that drive progression and aggressiveness of neuroblastoma
tumors, using gene expression data from cell lines isolated from high
risk neuroblastoma patients.

\hypertarget{specific}{%
\subsubsection{2.2. Specific}\label{specific}}

To investigate the nature of hypoxia signaling in neuroblastoma cells:
positive and negative outcomes in the discovery and validation
neuroblastoma cohorts, by means of the TET1 induction of 5-hmC
deposition triggered by hypoxia in the ADRN to MES transition in
neuroblastoma . Correlate neuroblastoma phenotypes hypoxia, HIF1A
expression, T-cell exhaustion, to the high risk variable, survival (OS
and/or EFS) and risk Hazards Ratio (Cox analysis). Write a machine
learning algorithm to evaluate the effect of hypoxia in survival of high
risk neuroblastoma patients profiled by 5-hmC deposition in their TE
elements in cfDNA.

\hypertarget{materials-and-methods}{%
\subsection{3. MATERIALS AND METHODS}\label{materials-and-methods}}

General pipeline for the machine learning algorithm development was
based on figure 1 of Rabiei et al. (2022), Nemade and Fegade (2023) and
Khalid et al. (2021). For an integration of deep learning on cancer
diagnosis, prognosis and treatment selection, please see Tran et al.
(2021). Another highly cited feature selection paper was Akay (2009).

\hypertarget{phenotype-quantification}{%
\subsubsection{3.1. Phenotype
quantification}\label{phenotype-quantification}}

Eight gene sets were used to characterize the impact of hypoxia on high
risk, survival, hazard ratio and cumulative hazard ratio.. , to quantify
the correlation between gene set signatures in tumors.

These algorithms were developed in precision oncology to identify
processes driving progression and aggressiveness of neuroblastoma
tumors, using gene expression data from cell lines isolated from high
risk neuroblastoma tumors.

\hypertarget{phenotypes-contributing-to-the-risk-category}{%
\subsubsection{3.2. Phenotypes contributing to the risk
category}\label{phenotypes-contributing-to-the-risk-category}}

We will determine the most important variables ot predict high risk in
neuroblastoma (Sato et al. 2019). Although the down regulated genes had
a significant impact on the separability of the low risk and high risk
categories, the 635 gene set did not improve that ability significantly
(Figure 1). However, this gene set was significant in the survival
analysis of the high risk group after the multiple test correction
(Figure 6). The 635 gene set did not improve the sensitivity of the
method (Table X), XX \% with and XX \% without this group as a selected
feature.

\hypertarget{principal-component-analysis}{%
\subsubsection{3.3. Principal Component
Analysis}\label{principal-component-analysis}}

Impact of feature on label outcome was evaluates using the Scree Plot of
Principal Component Analysis. Feature selection was revised by Chellappa
and Turaga (2022). For high profile paper about risk prediction, see
Placido et al. (2023). For methods and feature selection, please see
Fakoor et al. (2013). Feature selection can help in the design of gene
expression panels for early cancer detection. Recent advances have made
it possible to use machine learning and feature selection for detection
and early detection of imperceptible cancers such as cancers with
elusive symptomes or tumors that have challenging diagnostic criteria
Metcalf (2024). Our aim is for this framework to be used as a feature
selection tool for oncologic panels for diagnostic selection criteria as
in Costa et al. (2023).

\hypertarget{correlation-matrix-to-eliminate-redundant-features}{%
\subsubsection{3.4. Correlation matrix to eliminate redundant
features:}\label{correlation-matrix-to-eliminate-redundant-features}}

ADRN Norm vs Hypo Down (635) gene set was removed because the Scree plot
showed a high impact of the other decreasing proliferation gene sets.
Therefore in this part several pipelines can be proposed on how to
choose a threshold of the correlation to remove features that do not
increase specificity of the label of the sample.

\hypertarget{development-of-machine-learning-framework-for-risk-classification-and-diagnosis}{%
\subsubsection{3.4. Development of machine learning framework for risk
classification and
diagnosis}\label{development-of-machine-learning-framework-for-risk-classification-and-diagnosis}}

Development of a framework may help the implementation of liquid biopsy
protocols for management and therapeutics of oncologic patients. To
determine parameters contributing to neuroblastoma risk classification,
we constructed a platform for visualization of phenotypes driving tumor
progression using R version 4.1.1. (\url{http://www.r-project.org}) and
the Shiny and Caret packages. The first step for the platform
construction was to establish a algorithms for supervised learning and
classifiers as well as their hyperparameters. Algorithms tested included
Logistic Regression, Random Forest, Neural Network, Supporting Vector
Machines and Deep Learning were used to construct a non-linear
classification model.

Variable importance, correlation plots and PCA components will be
visualized using the Shiny package in R.

\hypertarget{results}{%
\subsection{4. RESULTS}\label{results}}

\hypertarget{principal-component-analysis-of-features-contributing-to-survival-and-high-risk-labels}{%
\subsubsection{4.1. Principal Component Analysis of features
contributing to survival and high risk
labels}\label{principal-component-analysis-of-features-contributing-to-survival-and-high-risk-labels}}

In this part we used PCA to determine which features are most relevant
to risk and or survival. Which features contribute the most to the risk
category and the MYCN category. After that, we will plot some
correlations between the features, for example inflammation and hypoxia.
Then there can be a plot comparing the scores of the phenotype comparing
the high and non-high-risk categories. Finally, we can draw a couple of
survival analysis to compare the scores of the phenotypes. It is
surprising that in the plot of Figure 1, we did not have an effect of
hypoxia down-regulation in determining an increase in the specificity of
the machine learning algorithm to separate the low risk and high risk
categories (Figure 1).

\begin{figure}
\centering
\includegraphics[width=1\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/cumulative_variance_pcs.png}
\caption{Cumulative variance and combined principal components to
evaluate the role of 8 gene set phenotypes in differentiating the
neuroblastoma high risk group.}
\end{figure}

\hypertarget{high-risk-neuroblastoma-is-characterized-by-increased-quiescence-and-low-inflammatory-activities-correlated-with-hypoxia-markers}{%
\subsubsection{4.2. High risk neuroblastoma is characterized by
increased quiescence and low inflammatory activities correlated with
hypoxia
markers}\label{high-risk-neuroblastoma-is-characterized-by-increased-quiescence-and-low-inflammatory-activities-correlated-with-hypoxia-markers}}

High risk classification of neuroblastoma represents the better
prognostication system available for the disease. Patients with less
than 50\% 5 years event-free survival are classified as high risk
patients (Cohn et al., 2009). To evaluate the correlation of hypoxia
phenotype scoring with high-risk neuroblastoma, we evaluated the hypoxia
gene-sets comparing the low and high risk groups. High risk patients
showed a trend to have increased scores for hypoxia phenotypes that
decrease cellular proliferation evaluated in the Discovery cohort
(Figure 2).

Considering the dual phenotypic effects of hypoxia, one in the genes
that are activated by the hypoxia signal and the other on genes that are
down-regulaeted, we evaluated the correlation of inflammation phenotypes
with the high-risk variable. The high-risk group presented decreased
inflammatory response events than the low risk group (Figure 2).
Together these results show a trend in which hypoxia modulates genes by
activation and inhibition of gene expression (Figure 2).
Hypoxia-activated genes seem to activate inflammation whereas
hypoxia-inhibited genes seem to decrease cell-cycle proliferation
(Figure 2).

\begin{figure}
\centering
\includegraphics[width=1\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/gene_sets_hr.png}
\caption{Correlation between cell cycle phenotypes impacted by hypoxia
and the high risk variable. A) HIF1A and B) HIF1A gene targets and C)
T-cell exhaustion and D) quiescence as defined in Cheung (2019) were all
associated with higher scores in high risk patients in the discovery
cohort with similar trends observed in the validation cohort.
Correlation between inflammatory phenotypes impacted by hypoxia and the
high risk variable. A) Hallmark inflammatory response, B) hallmark
hypoxia, C) quiescence as defined in Cheung 2019 and D) ST8A1 were all
phenotypes associated with higher scores in high risk patients. Data
shown is for the discovery cohort and similar trends were observed in
the validation cohort.}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=0.4\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/ggplot_hif1a.png}
\caption{Correlation between cell cycle phenotypes impacted by hypoxia
and the high risk variable. A) HIF1A and B) HIF1A gene targets and C)
T-cell exhaustion and D) quiescence as defined in Cheung (2019) were all
associated with higher scores in high risk patients in the discovery
cohort with similar trends observed in the validation cohort.}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=0.8\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/ggplot_inflammation.png}
\caption{Inflammation.}
\end{figure}

Now, I will plot the combined plot:

\begin{figure}
\centering
\includegraphics[width=0.9\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/combined.png}
\caption{Inflammation.}
\end{figure}

Now, I will plot ADRN Norm vs Hypo Down (635) plot:

\begin{figure}
\centering
\includegraphics[width=0.6\textwidth,height=\textheight]{../../../ReComBio Book English/recombio bookdown/figures/plot_fit_os_ADRN_Norm_vs_Hypo_Down_635.png}
\caption{Survival analysis of ADRN Norm vs Hypo Down (635) after
multiple test correction.}
\end{figure}

\hypertarget{predictive-accuracy-for-each-neuroblastoma-high-risk-classifier}{%
\subsubsection{4.3. Predictive accuracy for each neuroblastoma high risk
classifier}\label{predictive-accuracy-for-each-neuroblastoma-high-risk-classifier}}

Table X shows the predictive accuracy and area under the curve of each
classifier using different types of machine learning algorithms Sato et
al. (2019). We started with Linear regression model and found a Receiver
Operating Characteristic (ROC) Curve of 91\%. Sensitivity and
Specificity were 88 and 78\% respectively when
createDataPartition(df3\_nb\$high\_risk, times = 1, p = 0.67, list =
FALSE).

Table X.: ROC, Sensitivity and Specificity for the Logistic Regression
Model. ROC Sens Spec\\
0.919246 0.8895238 0.777381

\hypertarget{references}{%
\subsection{References}\label{references}}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-Akay2009}{}}%
Akay, Mehmet Fatih. 2009. \emph{Support Vector Machines Combined with
Feature Selection for Breast Cancer Diagnosis}. \emph{Expert Systems
with Applications}.

\leavevmode\vadjust pre{\hypertarget{ref-Chellappa2020}{}}%
Chellappa, Rama, and Pavan Turaga. 2022. \emph{Feature Selection}.
\emph{Computer Vision}.

\leavevmode\vadjust pre{\hypertarget{ref-Costa2023}{}}%
Costa, Maria Claudia, Teresa Maria Rosaria Noviello, Michele Ceccarelli,
and Luigi Cerulo. 2023. \emph{Robust Feature Selection Strategy Detects
a Panel of microRNAs as Putative Diagnostic Biomarkers in Breast
Cancer}. \emph{Proceedings of the 18th Conference on Computational
Intelligence Methods for Bioinformatics \& Biostatistics (CIBB 2023)}.

\leavevmode\vadjust pre{\hypertarget{ref-Fakoor2013}{}}%
Fakoor, Rasool, Faisal Ladhak, Azade Nazi, and Manfred Huber. 2013.
\emph{Using Deep Learning to Enhance Cancer Diagnosis and
Classification}. \emph{Proceedings of the 30th International Conference
on Ma- Chine Learning}.

\leavevmode\vadjust pre{\hypertarget{ref-Khalid2023}{}}%
Khalid, Arslan, Arif Mehmood, Amerah Alabrah, Bader Fahad Alkhamees,
Farhan Amin, Hussain AlSalman, and Gyu Sang Choi. 2021. \emph{Breast
Cancer Detection and Prevention Using Machine Learning}. \emph{Genome
Medicine}.

\leavevmode\vadjust pre{\hypertarget{ref-Metcalf2024}{}}%
Metcalf, Gavin. 2024. \emph{MicroRNAs: Circulating Biomarkers for the
Early Detection of Imperceptible Cancers via Biosensor and
Machine-Learning Advances}. \emph{Oncogene}.

\leavevmode\vadjust pre{\hypertarget{ref-Nemade2023}{}}%
Nemade, Varsha, and Vishal Fegade. 2023. \emph{Machine Learning
Techniques for Breast Cancer Prediction}. \emph{Procedia Computer
Science}.
\url{https://www.sciencedirect.com/science/article/pii/S1877050923001102}.

\leavevmode\vadjust pre{\hypertarget{ref-Placido2023}{}}%
Placido, Davide, Amalie D. Haue, Renato Umeton, Alexandra Franz, Lauren
Brais, Elizabeth Andrews, Debora S. Marks, et al. 2023. \emph{A Deep
Learning Algorithm to Predict Risk of Pancreatic Cancer from Disease
Trajectories}. \emph{Computer Vision}.

\leavevmode\vadjust pre{\hypertarget{ref-Rabiei2022}{}}%
Rabiei, Reza, Seyed Mohammad Ayyoubzadeh, Solmaz Sohrabei, Marzieh
Esmaeili, and Alireza Atashi. 2022. \emph{Prediction of Breast Cancer
Using Machine Learning Approaches}. \emph{J Biomed Phys Eng}.

\leavevmode\vadjust pre{\hypertarget{ref-Sato2019}{}}%
Sato, Masaya, Kentaro Morimoto, Shigeki Kajihara, and YutakaYatomi.
2019. \emph{Machine-Learning Approach for the Development of a Novel
Predictive Model for the Diagnosis of Hepatocellular Carcinoma}.
\emph{Scientific Reports}.
\url{https://www.scielo.br/j/icse/a/WC7GKD4py6Cq7cLdRvDZx3H/?lang=pt}.

\leavevmode\vadjust pre{\hypertarget{ref-Tran2021}{}}%
Tran, Khoa A., Olga Kondrashova, Andrew Bradley, Elizabeth D. Williams,
John V. Pearson, and Nicola Waddell. 2021. \emph{Deep Learning in Cancer
Diagnosis, Prognosis and Treatment Selection}. \emph{Genome Medicine}.

\end{CSLReferences}

\end{document}
