---
  title: "Genotype Information Extraction"
  author: "Gepoliano Chaves"
  date: "June 24th, 2024"
  output:
    html_document:
      df_print: paged
      toc: true
  bibliography: references.bib
---

\

## Summary

\

In this activity we process VCF files to generate a genotype file to be used in the machine learning framework.

\


This simple for loop looks into all vcf file to look for mutation:

```{bash, eval=F}
for vcf_file in *_snps.vcf; do 
  echo $vcf_file
  for position in 16069 16111; do
    echo $position
    grep --color $position $vcf_file; done
    done
```

\

This script is more complex and starts to print all information that I want, that variant is in file or not in file:

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  echo $vcf_file is the vcf file
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    echo $position is the position
    if
      grep --color -q -w $position $vcf_file
    then
      echo "position in file"
    else
      echo "mutation is not in file"
    fi
  done;
done
```

\

In this chunk I will register the reference if it is present in the vcf file.:

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  echo $vcf_file is the vcf file
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    echo $position is the position
    if
      grep --color -q -w $position $vcf_file
    then
      echo "position in file"
      ## Note that it has to be the vcf file variable in front of $position:
      ref=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      echo $ref is the reference alele
    else
      echo "mutation is not in file"
    fi
  done;
done
```

\

And then we also register if the alteration is present in the vcf file:

\

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  echo $vcf_file "xxxxxxxxxxxxxxxxxxxxxxxxxxxxx is the vcf file xxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    echo $position is the position
    if
      grep --color -q -w $position $vcf_file
    then
      echo "...................................position in file..................................."
      ## Note that it has to be the vcf file variable in front of $position:
      ref=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      ## Note that it also has to be the vcf file variable in front of $position:
      alt=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      echo $ref is the reference alele and $alt is the alteration
    else
      echo "mutation is not in file" $ref is the reference and  $alt is the alteration
    fi
  done;
done
```

\

Include position and genotype information

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  echo $vcf_file "................................... is the vcf file ..................................."
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    echo $position is the position
    if
      grep --color -q -w $position $vcf_file
    then ## Mutation is present
      echo "xxxxxxxxxxxxxxxxxxxxxxxxxxxxx position in file xxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
      ## Note that it has to be the vcf file variable in front of $position:
      ref=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      ## Note that it also has to be the vcf file variable in front of $position:
      alt=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      echo $ref is the reference alele and $alt is the alteration
      ## MT_position_bp Ref Alt VCF
      echo  "MT_"$position"_bp" $ref $alt "  then is  " $vcf_file
    else ## Mutation is NOT present
      echo "mutation is not in file" $ref is the reference and  $alt is the NOT alteration
      ## MT_position_bp Ref Alt VCF
      echo  "MT_"$position"_bp" $ref $ref "  then is  " $vcf_file
    fi
  done;
done
```

\

This is more complex and prints almost all information that I want. The last column is the name of the vcf file. There is a column in the middle that is the genotype column:

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  #echo $vcf_file
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    if
      grep --color -q -w $position $vcf_file
    then
      ref=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      alt=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      echo  "MT_"$position"bp" $i"    "$ref"1"$alt "  then is  " $vcf_file
    else
      echo  "MT_"$position"bp" $i"    "$ref"0"$ref "  else is  " $vcf_file
    fi
  done;
done
```

\

This script now prints variant name, the genotype of the variant and file name:

```{bash, eval=F, include=F}
cd /hb/home/gchaves/results/geraldo/mitochondria/snps_vcf
for vcf_file in *_snps.vcf; do 
  for position in 16069 16086 16093 16111 16124; do
    if
      grep --color -q -w $position $vcf_file
    then
      ref=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      alt=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      echo  "MT_"$position"bp" $i"    "$ref"1"$alt "  then is  " $vcf_file
    else
      echo  "MT_"$position"bp" $i"    "$ref"0"$ref "  else is  " $vcf_file
    fi
  done;
done | awk -F ' ' '{print $1, $3, $6}'
```

\

### Extract Genotypes

\

For these positions, it was yielding an error, that I found the solution to fix:

```{bash, eval=F}
cd /hb/home/gchaves/results/geraldo/mitochondria/snps_vcf
for vcf_file in *_snps.vcf; do 
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    if
      grep --color -q -w $position $vcf_file
    then
      ref=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      alt=$(grep -w $position $i | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      echo  "MT_"$position"bp" $i"    "$ref"1"$alt "  then is  " $vcf_file
    else
      echo  "MT_"$position"bp" $i"    "$ref"0"$ref "  else is  " $vcf_file
    fi
  done;
done | awk -F ' ' '{print $1, $3, $6}' | sed 's/T0/0/g' | sed 's/0T/0/g' | sed 's/T1C/1/g' > genotypes3.txt
```

\

This seemed to be working.

```{bash, eval=F, include=F}
for vcf_file in *_snps.vcf; do 
  # echo $vcf_file "................................... is the vcf file ..................................."
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    #echo $position is the position
    if
      grep --color -q -w $position $vcf_file
    then ## Mutation is present
     # echo "xxxxxxxxxxxxxxxxxxxxxxxxxxxxx position in file xxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
      ## Note that it has to be the vcf file variable in front of $position:
      ref=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $3}') 
      ## Note that it also has to be the vcf file variable in front of $position:
      alt=$(grep -w $position $vcf_file | awk '{print $1, $2, $4, $5}'  | awk -F ',' '{print $1}' | awk '{print $4}') 
      #echo $ref is the reference alele and $alt is the alteration
      ## MT_position_bp Ref Alt VCF
      echo  "MT_"$position"_bp" $ref $alt "  then is  " $vcf_file
    else ## Mutation is NOT present
     # echo "mutation is not in file" $ref is the reference and  $alt is the NOT alteration
      ## MT_position_bp Ref Alt VCF
      echo  "MT_"$position"_bp" $ref $ref "  then is  " $vcf_file
    fi
  done;
done | awk -F ' ' '{print $1, $2, $3, $6}' | sed 's/T T AF/0 AF/g' | sed 's/C C AF/0 AF/g' | sed 's/G G AF/0 AF/g' | sed 's/A A AF/0 AF/g'| sed 's/T C AF/1 AF/g' | sed 's/C T AF/1 AF/g' | sed 's/G A AF/1 AF/g'  > genotypes4.txt
```


### Transfer 

\

```{bash, eval=F}
scp gchaves@hb.ucsc.edu:/hb/home/gchaves/results/geraldo/mitochondria/snps_vcf/genotypes4.txt \
/Users/gepolianochaves/Desktop/Gepoliano/ReComBio\ Scientific/geraldo/data/genotypes4.txt
```

\

## References

\

<div id="refs"></div>

\

## Session Info

\

```{r session}
sessionInfo()
```