<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Gepoliano Chaves" />


<title>Genotype Information Extraction</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Gerenciamento Genômico e Algorítmico de Fenótipos, Riscos, Diagnósticos e Prognósticos em Dados de Doenças em Brasileiros</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="quokka-bilby.html">
    <span class="fa fa-solid fa-circle-info"></span>
     
    Sobre nós
  </a>
</li>
<li>
  <a href="codigo.html">
    <span class="fa fa-solid fa-laptop-file"></span>
     
    Códigos
  </a>
</li>
<li>
  <a href="PmD.html">
    <span class="fa fa-solid fa-people-arrows"></span>
     
    PmD
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Genotype Information Extraction</h1>
<h4 class="author">Gepoliano Chaves</h4>
<h4 class="date">June 24th, 2024</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#summary" id="toc-summary">Summary</a>
<ul>
<li><a href="#extract-genotypes" id="toc-extract-genotypes">Extract
Genotypes</a></li>
<li><a href="#transfer" id="toc-transfer">Transfer</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
<li><a href="#session-info" id="toc-session-info">Session Info</a></li>
</ul>
</div>

<p><br />
</p>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p><br />
</p>
<p>In this activity we process VCF files to generate a genotype file to
be used in the machine learning framework.</p>
<p><br />
</p>
<p>This simple for loop looks into all vcf file to look for
mutation:</p>
<pre class="bash"><code>for vcf_file in *_snps.vcf; do 
  echo $vcf_file
  for position in 16069 16111; do
    echo $position
    grep --color $position $vcf_file; done
    done</code></pre>
<p><br />
</p>
<p>This script is more complex and starts to print all information that
I want, that variant is in file or not in file:</p>
<p><br />
</p>
<p>In this chunk I will register the reference if it is present in the
vcf file.:</p>
<p><br />
</p>
<p>And then we also register if the alteration is present in the vcf
file:</p>
<p><br />
</p>
<p><br />
</p>
<p>Include position and genotype information</p>
<p><br />
</p>
<p>This is more complex and prints almost all information that I want.
The last column is the name of the vcf file. There is a column in the
middle that is the genotype column:</p>
<p><br />
</p>
<p>This script now prints variant name, the genotype of the variant and
file name:</p>
<p><br />
</p>
<div id="extract-genotypes" class="section level3">
<h3>Extract Genotypes</h3>
<p><br />
</p>
<p>For these positions, it was yielding an error, that I found the
solution to fix:</p>
<pre class="bash"><code>cd /hb/home/gchaves/results/geraldo/mitochondria/snps_vcf
for vcf_file in *_snps.vcf; do 
  for position in 16069 16086 16093 16111 16124 16126 16129 16144 16153 16172 16209 16222 16270 16294 16312 16327 16362; do
    if
      grep --color -q -w $position $vcf_file
    then
      ref=$(grep -w $position $i | awk &#39;{print $1, $2, $4, $5}&#39;  | awk -F &#39;,&#39; &#39;{print $1}&#39; | awk &#39;{print $3}&#39;) 
      alt=$(grep -w $position $i | awk &#39;{print $1, $2, $4, $5}&#39;  | awk -F &#39;,&#39; &#39;{print $1}&#39; | awk &#39;{print $4}&#39;) 
      echo  &quot;MT_&quot;$position&quot;bp&quot; $i&quot;    &quot;$ref&quot;1&quot;$alt &quot;  then is  &quot; $vcf_file
    else
      echo  &quot;MT_&quot;$position&quot;bp&quot; $i&quot;    &quot;$ref&quot;0&quot;$ref &quot;  else is  &quot; $vcf_file
    fi
  done;
done | awk -F &#39; &#39; &#39;{print $1, $3, $6}&#39; | sed &#39;s/T0/0/g&#39; | sed &#39;s/0T/0/g&#39; | sed &#39;s/T1C/1/g&#39; &gt; genotypes3.txt</code></pre>
<p><br />
</p>
<p>This seemed to be working.</p>
</div>
<div id="transfer" class="section level3">
<h3>Transfer</h3>
<p><br />
</p>
<pre class="bash"><code>scp gchaves@hb.ucsc.edu:/hb/home/gchaves/results/geraldo/mitochondria/snps_vcf/genotypes4.txt \
/Users/gepolianochaves/Desktop/Gepoliano/ReComBio\ Scientific/geraldo/data/genotypes4.txt</code></pre>
<p><br />
</p>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p><br />
</p>
<div id="refs">

</div>
<p><br />
</p>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<p><br />
</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.4.1 (2024-06-14)
## Platform: aarch64-apple-darwin20
## Running under: macOS Big Sur 11.4
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## time zone: America/Chicago
## tzcode source: internal
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] tibble_3.2.1     stringr_1.5.1    dplyr_1.1.4      kableExtra_1.4.0
## [5] maditr_0.8.4     ape_5.8         
## 
## loaded via a namespace (and not attached):
##  [1] jsonlite_1.8.9    compiler_4.4.1    tidyselect_1.2.1  Rcpp_1.0.13      
##  [5] xml2_1.3.6        parallel_4.4.1    jquerylib_0.1.4   systemfonts_1.1.0
##  [9] scales_1.3.0      yaml_2.3.10       fastmap_1.2.0     lattice_0.22-6   
## [13] R6_2.5.1          generics_0.1.3    knitr_1.48        munsell_0.5.1    
## [17] svglite_2.1.3     bslib_0.8.0       pillar_1.9.0      rlang_1.1.4      
## [21] utf8_1.2.4        cachem_1.1.0      stringi_1.8.4     xfun_0.48        
## [25] sass_0.4.9        viridisLite_0.4.2 cli_3.6.3         magrittr_2.0.3   
## [29] digest_0.6.37     grid_4.4.1        rstudioapi_0.16.0 fontawesome_0.5.2
## [33] lifecycle_1.0.4   nlme_3.1-166      vctrs_0.6.5       evaluate_1.0.0   
## [37] glue_1.8.0        data.table_1.16.0 fansi_1.0.6       colorspace_2.1-1 
## [41] rmarkdown_2.28    pkgconfig_2.0.3   tools_4.4.1       htmltools_0.5.8.1</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
