<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Gepoliano Chaves" />


<title>Binary Logistic Regression</title>

<script src="site_libs/header-attrs-2.22/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/lumen.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Gerenciamento Genômico e Algorítmico de Fenótipos, Riscos, Diagnósticos e Prognósticos em Dados de Doenças em Brasileiros</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="about.html">
    <span class="fa fa-solid fa-circle-info"></span>
     
    Sobre nós
  </a>
</li>
<li>
  <a href="codigo.html">
    <span class="fa fa-solid fa-laptop-file"></span>
     
    Códigos
  </a>
</li>
<li>
  <a href="PmD.html">
    <span class="fa fa-solid fa-people-arrows"></span>
     
    PmD
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Binary Logistic Regression</h1>
<h4 class="author">Gepoliano Chaves</h4>
<h4 class="date">June 24th, 2024</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#summary" id="toc-summary">Summary</a>
<ul>
<li><a href="#load-data" id="toc-load-data">1. Load data</a>
<ul>
<li><a href="#running-time" id="toc-running-time">Running Time</a></li>
<li><a href="#finally-taking-a-look-into-the-target-variable"
id="toc-finally-taking-a-look-into-the-target-variable">Finally taking a
look into the target variable</a></li>
</ul></li>
<li><a href="#building-the-model" id="toc-building-the-model">2.
Building the Model</a>
<ul>
<li><a href="#interpreting-logistic-regression-output"
id="toc-interpreting-logistic-regression-output">Interpreting Logistic
Regression Output</a></li>
<li><a href="#predicting-dependent-variabley-in-test-dataset"
id="toc-predicting-dependent-variabley-in-test-dataset">Predicting
Dependent Variable(Y) in Test Dataset</a></li>
</ul></li>
<li><a href="#evaluating-logistic-regression-model"
id="toc-evaluating-logistic-regression-model">3. Evaluating Logistic
Regression Model</a>
<ul>
<li><a
href="#converting-probability-to-class-values-in-the-training-dataset"
id="toc-converting-probability-to-class-values-in-the-training-dataset">Converting
probability to class values in the training dataset</a></li>
<li><a
href="#training-dataset-converting-from-probability-to-class-values"
id="toc-training-dataset-converting-from-probability-to-class-values">Training
dataset converting from probability to class values</a></li>
<li><a href="#accuracy" id="toc-accuracy">3.1 Accuracy</a></li>
<li><a href="#misclassification-rate"
id="toc-misclassification-rate">3.2. Misclassification Rate</a></li>
<li><a href="#true-positive-rate-recall-or-sensitivity"
id="toc-true-positive-rate-recall-or-sensitivity">3.3. True Positive
Rate – Recall or Sensitivity</a></li>
<li><a href="#true-negative-rate" id="toc-true-negative-rate">3.4. True
Negative Rate</a></li>
<li><a href="#precision" id="toc-precision">3.5. Precision</a></li>
<li><a href="#calculating-f-score"
id="toc-calculating-f-score">Calculating F-Score</a></li>
</ul></li>
<li><a href="#roc-curve" id="toc-roc-curve">5. ROC Curve</a>
<ul>
<li><a href="#concordance" id="toc-concordance">Concordance</a></li>
</ul></li>
<li><a href="#confusion-matrix" id="toc-confusion-matrix">6. Confusion
matrix</a></li>
</ul></li>
<li><a href="#references" id="toc-references">References</a></li>
<li><a href="#session-info" id="toc-session-info">Session Info</a></li>
</ul>
</div>

<p><br />
</p>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p><br />
</p>
<p>This pipeline classifies the origin of the genotype based on the
genotype information extracted from VCF files using a binary logistic
regression model. In section 3.3 below, I discuss a potential approach
to improve the Sensitivity rates of this framework using the divide and
conquer approach. Because of the problem I faced running the for loop,
the genotypes were constructed appending the extraction of the vcf
files. This notebook is based on the tutorial from RBloggers (2020).</p>
<p><br />
</p>
<div id="load-data" class="section level2">
<h2>1. Load data</h2>
<p><br />
</p>
<div id="running-time" class="section level3">
<h3>Running Time</h3>
<p><br />
</p>
<p>The chunk below is used to calculate the time necessary to run this
notebook.</p>
<pre class="r"><code>start_time &lt;- Sys.time()</code></pre>
<p><br />
</p>
<p>Load libraries with functions necessary to input data for the present
analysis.</p>
<pre class="r"><code>library(expss)
library(dplyr)
library(kableExtra)
library(ggplot2)
library(caret)
library(pROC)</code></pre>
<p><br />
</p>
<p>Load data that was created by appending the genotypes from the vcf
files.</p>
<pre class="r"><code>mt_features_df &lt;- read.table(&quot;./data/mt_dcast_appended_binary.txt&quot;, header = T)
dim(mt_features_df)</code></pre>
<pre><code>## [1] 34  6</code></pre>
<p><br />
</p>
<p>Look at features in the mitochondria genotype object. Each column is
a feature. In R, columns are called variables. In R, a table can be
called a data-frame.</p>
<pre class="r"><code>kable(mt_features_df, caption=&quot;Mitochondrial Features&quot;) %&gt;%
  kable_styling(&quot;striped&quot;, full_width = F, font_size = 12) %&gt;%
  scroll_box(width = &quot;100%&quot;, height = &quot;600px&quot;)</code></pre>
<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:600px; overflow-x: scroll; width:100%; ">
<table class="table table-striped" style="font-size: 12px; width: auto !important; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">
Mitochondrial Features
</caption>
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
fastaID
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
origin
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
MT_16069bp
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
MT_16086bp
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
MT_16093bp
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
MT_16111bp
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
AF243627
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243628
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243629
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243630
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243631
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243632
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243633
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243700
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243701
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243702
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243703
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243704
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243705
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243706
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243707
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243708
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243709
</td>
<td style="text-align:left;">
Amerindian
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243780
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243781
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243782
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243783
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243784
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243785
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243786
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243787
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243788
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243789
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243790
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243791
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243792
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243793
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243794
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243795
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
AF243796
</td>
<td style="text-align:left;">
European
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
<p><br />
</p>
<p>Make origin a factor</p>
<pre class="r"><code>mt_features_df$origin &lt;- as.factor(mt_features_df$origin)
# mt_features_df &lt;- mt_features_df %&gt;% select(-c(&quot;GenBankID&quot;))
mt_features_df &lt;- mt_features_df %&gt;% select(-c(&quot;fastaID&quot;))</code></pre>
<p><br />
</p>
</div>
<div id="finally-taking-a-look-into-the-target-variable"
class="section level3">
<h3>Finally taking a look into the target variable</h3>
<p><br />
</p>
<pre class="r"><code>ggplot(mt_features_df, aes(MT_16069bp)) +
  geom_histogram(aes(fill = origin), color = &quot;black&quot;, binwidth = 0.1)</code></pre>
<p><img src="05-binary-logistic-regression-appended_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p><br />
</p>
</div>
</div>
<div id="building-the-model" class="section level2">
<h2>2. Building the Model</h2>
<p><br />
</p>
<p>Let’s create a partition of the dataset. We are splitting the data
into train and test. We are partitioning 60% of the samples to the train
group and 40% to the test group.</p>
<p>To solve the genotype error, I read the <span
class="citation">Stackoverflow (2021)</span> reference. To deal with
levels that I was not familiar in the dataset, I read the <span
class="citation">Stackoverflow (2014)</span> reference.</p>
<pre class="r"><code>set.seed(2015)
index &lt;- createDataPartition(mt_features_df$origin, p = .60, list = FALSE)
class(index)</code></pre>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot;</code></pre>
<pre class="r"><code>index</code></pre>
<pre><code>##       Resample1
##  [1,]         1
##  [2,]         2
##  [3,]         4
##  [4,]         6
##  [5,]         7
##  [6,]         8
##  [7,]         9
##  [8,]        10
##  [9,]        11
## [10,]        12
## [11,]        14
## [12,]        18
## [13,]        19
## [14,]        20
## [15,]        21
## [16,]        25
## [17,]        26
## [18,]        27
## [19,]        28
## [20,]        31
## [21,]        32
## [22,]        33</code></pre>
<pre class="r"><code>train_mt &lt;- mt_features_df[index, ]
train_mt</code></pre>
<pre><code>##        origin MT_16069bp MT_16086bp MT_16093bp MT_16111bp
## 1  Amerindian          0          0          0          1
## 2  Amerindian          0          0          0          0
## 4  Amerindian          0          0          0          0
## 6  Amerindian          0          0          0          0
## 7  Amerindian          0          0          0          0
## 8  Amerindian          0          0          0          1
## 9  Amerindian          0          0          0          0
## 10 Amerindian          0          0          0          0
## 11 Amerindian          0          0          0          0
## 12 Amerindian          0          0          0          0
## 14 Amerindian          0          0          0          1
## 18   European          0          0          0          0
## 19   European          0          0          0          0
## 20   European          0          0          0          0
## 21   European          0          0          0          0
## 25   European          0          0          0          0
## 26   European          0          0          0          0
## 27   European          0          0          1          0
## 28   European          0          0          0          0
## 31   European          0          0          0          0
## 32   European          0          0          0          0
## 33   European          1          0          0          0</code></pre>
<p><br />
</p>
<p>Look at counts per category</p>
<pre class="r"><code>cross_cases(train_mt, origin)</code></pre>
<table class="gmisc_table" style="border-collapse: collapse; margin-top: 1em; margin-bottom: 1em;">
<thead>
<tr>
<th style="border-bottom: 1px solid grey; font-weight: 900; border-top: 2px solid grey; text-align: center;">
</th>
<th style="font-weight: 900; border-bottom: 1px solid grey; border-top: 2px solid grey; text-align: center;">
 #Total 
</th>
</tr>
</thead>
<tbody>
<tr>
<td colspan="2" style="font-weight: 900;">
 origin 
</td>
</tr>
<tr>
<td style="text-align: left;">
   Amerindian 
</td>
<td style="text-align: right;">
11
</td>
</tr>
<tr>
<td style="text-align: left;">
   European 
</td>
<td style="text-align: right;">
11
</td>
</tr>
<tr>
<td style="border-bottom: 2px solid grey; text-align: left;">
   #Total cases 
</td>
<td style="border-bottom: 2px solid grey; text-align: right;">
22
</td>
</tr>
</tbody>
</table>
<p><br />
</p>
<p>Build test dataset</p>
<pre class="r"><code>test_mt &lt;- mt_features_df[-index, ]
test_mt</code></pre>
<pre><code>##        origin MT_16069bp MT_16086bp MT_16093bp MT_16111bp
## 3  Amerindian          0          0          0          0
## 5  Amerindian          0          0          0          0
## 13 Amerindian          0          0          0          0
## 15 Amerindian          0          0          0          0
## 16 Amerindian          1          0          0          0
## 17 Amerindian          0          0          0          0
## 22   European          0          0          0          0
## 23   European          1          0          0          0
## 24   European          0          0          0          0
## 29   European          0          1          0          0
## 30   European          0          0          0          0
## 34   European          0          0          0          0</code></pre>
<p><br />
</p>
<p>Training the model</p>
<pre class="r"><code>logistic_model_mt &lt;- glm(origin ~ ., family = binomial(), train_mt)
# mod2$xlevels[[&quot;y&quot;]] &lt;- union(mod2$xlevels[[&quot;y&quot;]], levels(test$y))</code></pre>
<p><br />
</p>
<p>Checking the model</p>
<pre class="r"><code>summary(logistic_model_mt)</code></pre>
<pre><code>## 
## Call:
## glm(formula = origin ~ ., family = binomial(), data = train_mt)
## 
## Deviance Residuals: 
##    Min      1Q  Median      3Q     Max  
## -1.228  -1.228   0.000   1.128   1.128  
## 
## Coefficients: (1 not defined because of singularities)
##              Estimate Std. Error z value Pr(&gt;|z|)
## (Intercept)    0.1178     0.4859   0.242    0.808
## MT_16069bp    18.4483  6522.6386   0.003    0.998
## MT_16086bp         NA         NA      NA       NA
## MT_16093bp    18.4483  6522.6386   0.003    0.998
## MT_16111bp   -18.6839  3765.8472  -0.005    0.996
## 
## (Dispersion parameter for binomial family taken to be 1)
## 
##     Null deviance: 30.498  on 21  degrees of freedom
## Residual deviance: 23.508  on 18  degrees of freedom
## AIC: 31.508
## 
## Number of Fisher Scoring iterations: 17</code></pre>
<p><br />
</p>
<div id="interpreting-logistic-regression-output"
class="section level3">
<h3>Interpreting Logistic Regression Output</h3>
<p><br />
</p>
<pre class="r"><code>Sigmoid function, p = exp(-0.591532)/(1+exp(-0.591532))</code></pre>
<p><br />
</p>
</div>
<div id="predicting-dependent-variabley-in-test-dataset"
class="section level3">
<h3>Predicting Dependent Variable(Y) in Test Dataset</h3>
<p><br />
</p>
<pre class="r"><code># Predicting in the test dataset
pred_prob_mt &lt;- predict(logistic_model_mt, test_mt, type=&quot;response&quot;)</code></pre>
<pre><code>## Warning in predict.lm(object, newdata, se.fit, scale = 1, type = if (type == :
## prediction from a rank-deficient fit may be misleading</code></pre>
<pre class="r"><code>pred_prob_mt</code></pre>
<pre><code>##         3         5        13        15        16        17        22        23 
## 0.5294118 0.5294118 0.5294118 0.5294118 1.0000000 0.5294118 0.5294118 1.0000000 
##        24        29        30        34 
## 0.5294118 0.5294118 0.5294118 0.5294118</code></pre>
<p><br />
</p>
<div id="constructing-the-predictions-dataframe" class="section level4">
<h4>Constructing the predictions dataframe</h4>
<ul>
<li><p>This is based on notebook confusion-table.Rmd</p></li>
<li><p>pred_prob, from above, is a factor</p></li>
</ul>
<pre class="r"><code>predictions_mt &lt;- cbind(data.frame(train_preds=pred_prob_mt, 
                                test_mt$origin))
predictions_mt</code></pre>
<pre><code>##    train_preds test_mt.origin
## 3    0.5294118     Amerindian
## 5    0.5294118     Amerindian
## 13   0.5294118     Amerindian
## 15   0.5294118     Amerindian
## 16   1.0000000     Amerindian
## 17   0.5294118     Amerindian
## 22   0.5294118       European
## 23   1.0000000       European
## 24   0.5294118       European
## 29   0.5294118       European
## 30   0.5294118       European
## 34   0.5294118       European</code></pre>
<p>Then the confusion matrix can be calculated</p>
<pre class="r"><code>cm &lt;- caret::confusionMatrix(predictions_mt$train_preds, predictions_mt$test_mt.origin)</code></pre>
<p>This is the error</p>
<pre class="r"><code>Error in `match.arg()`:
! &#39;arg&#39; should be one of &quot;link&quot;, &quot;response&quot;, &quot;terms&quot;
Backtrace:
 1. stats::predict(logistic_model_mt, test_mt, type = &quot;raw&quot;)
 2. stats::predict.glm(logistic_model_mt, test_mt, type = &quot;raw&quot;)
 3. base::match.arg(type)
Execution halted</code></pre>
<p><br />
</p>
</div>
</div>
</div>
<div id="evaluating-logistic-regression-model" class="section level2">
<h2>3. Evaluating Logistic Regression Model</h2>
<p><br />
</p>
<div id="converting-probability-to-class-values-in-the-training-dataset"
class="section level3">
<h3>Converting probability to class values in the training dataset</h3>
<p><br />
</p>
<p>It does not make sense the labels to be America or Africa for this
notebook since here we are using currently, Amerindian and European
genotypes of mitochondrial Brazilian sequences. Therefore I am
commenting on that line that uses the probability value to set the
labels up. To update the labels I will make America the Amerindian label
and Africa the European label. My first impression was that the accuracy
increased significantly.</p>
<pre class="r"><code># Converting from probability to actual output
# train_mt$pred_class &lt;- ifelse(logistic_model_mt$fitted.values &gt;= 0.5, &quot;America&quot;, &quot;Africa&quot;)
train_mt$pred_class &lt;- ifelse(logistic_model_mt$fitted.values &gt;= 0.5, &quot;Amerindian&quot;, &quot;European&quot;)
dim(train_mt)</code></pre>
<pre><code>## [1] 22  6</code></pre>
<pre class="r"><code>train_mt</code></pre>
<pre><code>##        origin MT_16069bp MT_16086bp MT_16093bp MT_16111bp pred_class
## 1  Amerindian          0          0          0          1   European
## 2  Amerindian          0          0          0          0 Amerindian
## 4  Amerindian          0          0          0          0 Amerindian
## 6  Amerindian          0          0          0          0 Amerindian
## 7  Amerindian          0          0          0          0 Amerindian
## 8  Amerindian          0          0          0          1   European
## 9  Amerindian          0          0          0          0 Amerindian
## 10 Amerindian          0          0          0          0 Amerindian
## 11 Amerindian          0          0          0          0 Amerindian
## 12 Amerindian          0          0          0          0 Amerindian
## 14 Amerindian          0          0          0          1   European
## 18   European          0          0          0          0 Amerindian
## 19   European          0          0          0          0 Amerindian
## 20   European          0          0          0          0 Amerindian
## 21   European          0          0          0          0 Amerindian
## 25   European          0          0          0          0 Amerindian
## 26   European          0          0          0          0 Amerindian
## 27   European          0          0          1          0 Amerindian
## 28   European          0          0          0          0 Amerindian
## 31   European          0          0          0          0 Amerindian
## 32   European          0          0          0          0 Amerindian
## 33   European          1          0          0          0 Amerindian</code></pre>
<pre class="r"><code># Generating the classification table
ctab_train_mt &lt;- table(train_mt$origin, train_mt$pred_class)
ctab_train_mt</code></pre>
<pre><code>##             
##              Amerindian European
##   Amerindian          8        3
##   European           11        0</code></pre>
<p><br />
</p>
</div>
<div id="training-dataset-converting-from-probability-to-class-values"
class="section level3">
<h3>Training dataset converting from probability to class values</h3>
<p><br />
</p>
<pre class="r"><code># Converting from probability to actual output
# test_mt$pred_class &lt;- ifelse(pred_prob_mt &gt;= 0.5, &quot;America&quot;, &quot;Africa&quot;)
test_mt$pred_class &lt;- ifelse(pred_prob_mt &gt;= 0.5, &quot;Amerindian&quot;, &quot;European&quot;)

# Generating the classification table
ctab_test_mt &lt;- table(test_mt$origin, test_mt$pred_class)
ctab_test_mt</code></pre>
<pre><code>##             
##              Amerindian
##   Amerindian          6
##   European            6</code></pre>
<p><br />
</p>
</div>
<div id="accuracy" class="section level3">
<h3>3.1 Accuracy</h3>
<p><br />
</p>
<p><span class="math display">\[\begin{align*}

Accuracy &amp; = \frac{TP + TN}{TN + FP + FN + TP} \\

\end{align*}\]</span></p>
<p><br />
</p>
<p>Accuracy in Training dataset</p>
<pre class="r"><code>accuracy_train_mt &lt;- sum(diag(ctab_train_mt))/sum(ctab_train_mt)*100
accuracy_train_mt</code></pre>
<pre><code>## [1] 36.36364</code></pre>
<p><br />
</p>
<p>Accuracy in Test dataset</p>
<pre class="r"><code>accuracy_test_mt &lt;- sum(diag(ctab_test_mt))/sum(ctab_test_mt)*100
accuracy_test_mt</code></pre>
<pre><code>## [1] 50</code></pre>
<p><br />
</p>
<p>A model is considered fairly good if the model accuracy is greater
than 70%.</p>
<p><br />
</p>
</div>
<div id="misclassification-rate" class="section level3">
<h3>3.2. Misclassification Rate</h3>
<p><br />
</p>
<p><span class="math display">\[\begin{align*}

Misclassification  Rate &amp; = \frac{FP+FN}{TN + FP + FN + TP} \\

\end{align*}\]</span></p>
<p><br />
</p>
</div>
<div id="true-positive-rate-recall-or-sensitivity"
class="section level3">
<h3>3.3. True Positive Rate – Recall or Sensitivity</h3>
<p><br />
</p>
<p><span class="math display">\[\begin{align*}

Sensitivity &amp; = \frac{TP}{FN + TP} \\

\end{align*}\]</span></p>
<p><br />
</p>
<p>Recall in Train dataset</p>
<pre class="r"><code>Recall &lt;- (ctab_train_mt[2, 2]/sum(ctab_train_mt[2, ]))*100
Recall</code></pre>
<pre><code>## [1] 0</code></pre>
<pre class="r"><code>ctab_test_mt</code></pre>
<pre><code>##             
##              Amerindian
##   Amerindian          6
##   European            6</code></pre>
<p><br />
</p>
<p>Recall in Test dataset</p>
<pre class="r"><code>Recall_test &lt;- (ctab_test_mt[2, 2]/sum(ctab_test_mt[2, ]))*100
Recall_test</code></pre>
<p><br />
</p>
<p>One potential direction to improve sensitivity in the train and test
datasets is to use haplogroups instead of SNPs. Take for example the
Huntington’s disease (HD) diagnosis. A person is positive if they have
more than 35 CAG repeats. If for example we take a haplogroup of 50 bp,
we can classify people with 40 CAG repeats as positive whereas
individuals that have the haplogroup with less than 40 repeats are
negative for HD.</p>
<p>Using this approach we can divide the genome in smaller segments and
the algorithm will have to only work in these segments. We can
therefore, use the divide and conquer approach to diagnose the disease
using genotype information.</p>
<p><br />
</p>
</div>
<div id="true-negative-rate" class="section level3">
<h3>3.4. True Negative Rate</h3>
<p><br />
</p>
<p><span class="math display">\[\begin{align*}

TNR &amp; = \frac{TN}{TN + FP} \\

\end{align*}\]</span></p>
<p><br />
</p>
<p>TNR in Train dataset</p>
<pre class="r"><code>TNR &lt;- (ctab_train_mt[1, 1]/sum(ctab_train_mt[1, ]))*100
TNR</code></pre>
<pre><code>## [1] 72.72727</code></pre>
<p><br />
</p>
</div>
<div id="precision" class="section level3">
<h3>3.5. Precision</h3>
<p><br />
</p>
<p><span class="math display">\[\begin{align*}

TNR &amp; = \frac{TP}{FP + TP} \\

\end{align*}\]</span></p>
<p><br />
</p>
<pre class="r"><code># Precision = TP/FP + TP


# Precision in Train dataset
Precision &lt;- (ctab_train_mt[2, 2]/sum(ctab_train_mt[, 2]))*100
Precision</code></pre>
<pre><code>## [1] 0</code></pre>
<p><br />
</p>
</div>
<div id="calculating-f-score" class="section level3">
<h3>Calculating F-Score</h3>
<p><br />
</p>
<pre class="r"><code>F_Score &lt;- (2 * Precision * Recall / (Precision + Recall))/100
F_Score</code></pre>
<p><br />
</p>
</div>
</div>
<div id="roc-curve" class="section level2">
<h2>5. ROC Curve</h2>
<p><br />
</p>
<pre class="r"><code>roc_mt &lt;- roc(train_mt$origin, logistic_model_mt$fitted.values)</code></pre>
<pre><code>## Setting levels: control = Amerindian, case = European</code></pre>
<pre><code>## Setting direction: controls &lt; cases</code></pre>
<pre class="r"><code>roc_mt</code></pre>
<pre><code>## 
## Call:
## roc.default(response = train_mt$origin, predictor = logistic_model_mt$fitted.values)
## 
## Data: logistic_model_mt$fitted.values in 11 controls (train_mt$origin Amerindian) &lt; 11 cases (train_mt$origin European).
## Area under the curve: 0.7025</code></pre>
<pre class="r"><code>plot.roc(roc_mt)</code></pre>
<p><img src="05-binary-logistic-regression-appended_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<pre class="r"><code>auc(roc_mt)</code></pre>
<pre><code>## Area under the curve: 0.7025</code></pre>
<p><br />
</p>
<p>Note that per the <span class="citation">Wikipedia (2024)</span>
article, our classifier shows better results than random guess, which is
a good start.</p>
<p><br />
</p>
<div id="concordance" class="section level3">
<h3>Concordance</h3>
<p><br />
</p>
<pre class="r numberLines"><code>library(InformationValue)</code></pre>
<pre><code>## 
## Attaching package: &#39;InformationValue&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:caret&#39;:
## 
##     confusionMatrix, precision, sensitivity, specificity</code></pre>
<pre class="r numberLines"><code>Concordance(logistic_model_mt$y,logistic_model_mt$fitted.values)</code></pre>
<pre><code>## $Concordance
## [1] 0.4049587
## 
## $Discordance
## [1] 0.5950413
## 
## $Tied
## [1] 0
## 
## $Pairs
## [1] 121</code></pre>
<p><br />
</p>
</div>
</div>
<div id="confusion-matrix" class="section level2">
<h2>6. Confusion matrix</h2>
<p><br />
</p>
<p>In this reference for the Confusion Matrix according to <span
class="citation">CARET (2023)</span>, we can find the equations for
sensitivity and accuracy.</p>
<p><br />
</p>
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p><br />
</p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-CARET2023" class="csl-entry">
CARET. 2023. <em>confusionMatrix: Create a Confusion Matrix</em>. <a
href="https://rdrr.io/cran/caret/man/confusionMatrix.html">https://rdrr.io/cran/caret/man/confusionMatrix.html</a>.
</div>
<div id="ref-stackOverflow2014" class="csl-entry">
Stackoverflow. 2014. <em>"Factor Has New Levels" Error for Variable i’m
Not Using</em>. <a
href="https://stackoverflow.com/questions/22315394/factor-has-new-levels-error-for-variable-im-not-using">https://stackoverflow.com/questions/22315394/factor-has-new-levels-error-for-variable-im-not-using</a>.
</div>
<div id="ref-stackOverflow2021genotypes" class="csl-entry">
———. 2021. <a
href="https://stackoverflow.com/questions/65896181/glm-error-in-evalfamilyinitialize-y-values-must-be-0-y-1-but-values-a">https://stackoverflow.com/questions/65896181/glm-error-in-evalfamilyinitialize-y-values-must-be-0-y-1-but-values-a</a>.
</div>
<div id="ref-Wikipedia2024roc" class="csl-entry">
Wikipedia. 2024. <em>Receiver Operating Characteristic</em>. <a
href="https://en.wikipedia.org/wiki/Receiver_operating_characteristic">https://en.wikipedia.org/wiki/Receiver_operating_characteristic</a>.
</div>
</div>
<p><br />
</p>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<p><br />
</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] InformationValue_1.2.3 pROC_1.18.0            caret_6.0-94          
##  [4] lattice_0.22-6         expss_0.11.4           lubridate_1.9.2       
##  [7] forcats_1.0.0          purrr_1.0.1            readr_2.1.4           
## [10] tidyr_1.3.0            ggplot2_3.4.2          tidyverse_2.0.0       
## [13] microdatasus_1.4.4     data.table_1.15.2      reshape2_1.4.4        
## [16] tibble_3.2.1           stringr_1.5.0          dplyr_1.1.2           
## [19] kableExtra_1.3.4       maditr_0.8.4           ape_5.7-1             
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-162         matrixStats_1.1.0    fontawesome_0.5.2   
##  [4] webshot_0.5.5        httr_1.4.7           tools_4.1.1         
##  [7] backports_1.4.1      bslib_0.5.0          utf8_1.2.3          
## [10] R6_2.5.1             rpart_4.1.23         colorspace_2.1-0    
## [13] nnet_7.3-19          withr_3.0.0          tidyselect_1.2.0    
## [16] compiler_4.1.1       cli_3.6.2            rvest_1.0.3         
## [19] htmlTable_2.4.1      xml2_1.3.4           labeling_0.4.3      
## [22] sass_0.4.6           scales_1.2.1         checkmate_2.3.1     
## [25] systemfonts_1.0.4    digest_0.6.35        rmarkdown_2.22      
## [28] svglite_2.1.1        pkgconfig_2.0.3      htmltools_0.5.7     
## [31] parallelly_1.37.1    fastmap_1.1.1        highr_0.10          
## [34] htmlwidgets_1.6.2    rlang_1.1.3          rstudioapi_0.15.0   
## [37] jquerylib_0.1.4      farver_2.1.1         generics_0.1.3      
## [40] jsonlite_1.8.8       ModelMetrics_1.2.2.2 magrittr_2.0.3      
## [43] huxtable_5.5.2       Matrix_1.5-1         Rcpp_1.0.12         
## [46] munsell_0.5.0        fansi_1.0.6          lifecycle_1.0.4     
## [49] stringi_1.8.3        yaml_2.3.7           MASS_7.3-60.0.1     
## [52] plyr_1.8.8           recipes_1.0.6        grid_4.1.1          
## [55] listenv_0.9.1        parallel_4.1.1       crayon_1.5.2        
## [58] splines_4.1.1        hms_1.1.3            knitr_1.45          
## [61] pillar_1.9.0         stats4_4.1.1         future.apply_1.11.1 
## [64] codetools_0.2-19     glue_1.7.0           evaluate_0.23       
## [67] vctrs_0.6.2          tzdb_0.4.0           foreach_1.5.2       
## [70] gtable_0.3.4         future_1.33.1        assertthat_0.2.1    
## [73] cachem_1.0.8         xfun_0.42            gower_1.0.1         
## [76] prodlim_2023.03.31   class_7.3-22         survival_3.5-5      
## [79] viridisLite_0.4.2    timeDate_4032.109    iterators_1.0.14    
## [82] hardhat_1.3.1        lava_1.7.2.1         globals_0.16.3      
## [85] timechange_0.2.0     ipred_0.9-14</code></pre>
<p><br />
</p>
<pre class="r"><code>end_time &lt;- Sys.time()
end_time - start_time</code></pre>
<pre><code>## Time difference of 1.727061 secs</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
