---
title: "Merge mtDNA VCFs"

author: "Gepoliano Chaves"

date: "September 18th, 2024"

output:
  html_document:
    df_print: paged
    toc: true
---

\

### 1. Solve the scp transfer problem

\

This reference solved the rsync and scp problems:

https://serverfault.com/questions/304125/rsync-seems-incompatible-with-bashrc-causes-is-your-shell-clean

```{bash transfer fasta folder, eval=F}
rsync -avz ~/Desktop/Gepoliano/ReComBio Scientific/geraldo/data/mt_prs_capital.fasta gchaves@hb.ucsc.edu:/hb/home/gchaves/results/geraldo/prs
```

\

scp protocol is working again:

```{bash, eval=F}
scp -r ~/Desktop/Gepoliano/ReComBio\ Scientific/geraldo/data/mt_prs_capital_split_files gchaves@hb.ucsc.edu:/hb/home/gchaves/results/geraldo/prs
```


\

### 2. Variant Call Script for PRS of mtDNA:

\

This part is from file ~/scripts/align.sh :

```{bash, eval=F}
#!/bin/bash
#SBATCH --partition=128x24
#SBATCH --job-name=align_prs               # Job name
#SBATCH --mail-type=ALL                    # Mail events (NONE, BEGIN, END, FAIL, ALL)
#SBATCH --mail-user=gchaves@ucsc.edu       # Where to send mail	
#SBATCH --nodes=1                          # Use one node
#SBATCH --ntasks=1                         # Run a single task	
#SBATCH --cpus-per-task=4                  # Number of CPU cores per task
#SBATCH --output=align_prs.out             # Standard output and error log
# 
# module load gcc/5.4.0

SECONDS=0

source ~/.bashrc
echo "This is the task in the for loop"
REFERENCE=~/references/hg19/GRCh37.p13.genome.fa
cd /hb/home/gchaves/results/geraldo/prs/fasta

module load samtools
module load bcftools
module load bwa


## for loop specific to the individual fasta files
for fastaFile in AF*.fasta; do
  
  ## Pick just the eight firt letters of the string
  fastaFileName=${fastaFile:0:8}
  echo "started sample $fastaFileName"
  
  ## Align
  bwa mem -t 4 $REFERENCE $fastaFile > ~/results/geraldo/prs/sam/$fastaFileName".sam"
  echo "Just finished aligning sample $fastaFileName"
  
  ## SAM to BAM
  samtools view -S -b ~/results/geraldo/prs/sam/$fastaFileName".sam" > \
  ~/results/geraldo/prs/bam/$fastaFileName".bam"
  echo "Finished SAM to BAM of sample $fastaFileName"
  
  # Aligned as in Homer
  #samtools view -S -b test/AF243780.sam > test/AF243780.bam
  
  ## Samtools uses reference FASTA to detect "piles" in the alignment
  bcftools mpileup -f $REFERENCE ~/results/geraldo/prs/bam/$fastaFileName".bam" > \
  ~/results/geraldo/prs/bcf/$fastaFileName".bcf"
  
  ## Bcftools extracts SNPs
  echo "started Bcftools SNPs for sample $fastaFileName"
  bcftools view -v snps ~/results/geraldo/prs/bcf/$fastaFileName".bcf" > ~/results/geraldo/prs/snps_vcf/$fastaFileName"_snps.vcf"

done

echo "Finished entire for loop"

duration=$SECONDS

echo "$((duration / 60)) minutes and $((duration % 60)) seconds elapsed."
```

\

Create folder to bgzip files:

```{bash, eval=F}
mkdir bgzip_vcf
cp AF* bgzip_vcf
cd bgzip_vcf
for i in AF243*; do echo $i; done > files_list.txt
```


\

To load bgzip:
\

```{bash, eval=F}
module load htslib/1.17
```

Commands from cells-08-00962-s001.pdf:

Zip files, to comply with requirements of bcftools

```{bash, eval=F}
for x in $(cat files_list.txt); do bgzip $x; done
```

\

Change the list of files with .gz files:

\

```{bash, eval=F}
for i in AF243*; do echo $i; done > files_list.txt
```

\

Index files, to comply with requirements of bcftools

```{bash, eval=F}
for x in $(cat files_list.txt); do tabix $x; done
```

\

```{bash, eval=F}
module load bcftools
```

\

```{bash, eval=F}
bcftools merge --missing-to-ref --force-samples \
AF243627_snps.vcf.gz \
AF243628_snps.vcf.gz \
AF243629_snps.vcf.gz \
AF243634_snps.vcf.gz \
AF243635_snps.vcf.gz \
AF243636_snps.vcf.gz \
AF243637_snps.vcf.gz \
AF243638_snps.vcf.gz \
AF243639_snps.vcf.gz \
AF243640_snps.vcf.gz \
AF243700_snps.vcf.gz \
AF243791_snps.vcf.gz \
AF243792_snps.vcf.gz \
AF243793_snps.vcf.gz \
AF243794_snps.vcf.gz \
AF243795_snps.vcf.gz \
AF243796_snps.vcf.gz > mt_prs_merged.vcf
```

Use vcftools to extract ped and map from merged vcf file for plink downstream
analysis

```{bash, eval=F}
vcftools --vcf mt_prs_merged.vcf --out mt_prs_plink_output --plink
```

Use plink to run Fisherâ€™s exact test statistics on all SNPs contained in the ped
and map files

```{bash, eval=F}
plink --noweb --ped formatted_plink_output.ped --map
formatted_plink_output.map --fisher
```



